src/main/java/org/testng/DependencyMap.java

```diff
@@ -63,16 +61,32 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
       // If they are in the same class hierarchy, they must belong to the same instance,
       // otherwise, it's a method depending on a method in a different class so we
       // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
-      } else {
+      if (isSameInstance(fromMethod, m) || belongToDifferentClassHierarchy(fromMethod, m) || hasNoInstance(fromMethod, m)) {
         return m;
       }
     }
     throw new TestNGException("Method \"" + fromMethod
         + "\" depends on nonexistent method \"" + methodName + "\"");
   }
 
+  private static boolean belongToDifferentClassHierarchy(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    return ! baseClassMethod.getRealClass().isAssignableFrom(derivedClassMethod.getRealClass());
+  }
+
+  private static boolean hasNoInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null || baseInstance != null;
+  }
+
+  private static boolean isSameInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null &&
```

Comment:
Ok for me but I think we can merge `hasNoInstance` and `isSameInstance`. WDYT?

Reply:
@juherr - I have already tried doing that. It breaks either the PR changes (or) breaks tests that are part of  `org.testng.internal.DynamicGraphHelperTest`

So we will have to stick to this.